File: core/planner/src/main/java/ch/epfl/dias/calcite/adapter/pelago/rules/PelagoProjectPushBelowUnpack.java
package ch.epfl.dias.calcite.adapter.pelago.rules;

import ch.epfl.dias.calcite.adapter.pelago.*;
import org.apache.calcite.plan.RelOptRule;
import org.apache.calcite.plan.RelOptRuleCall;
import org.apache.calcite.plan.RelOptUtil;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.core.Project;
import org.apache.calcite.rex.RexInputRef;
import org.apache.calcite.rex.RexNode;
import org.apache.calcite.util.mapping.Mappings;

import com.google.common.collect.ImmutableMap;

import java.util.ArrayList;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;

public class PelagoProjectPushBelowUnpack extends RelOptRule {
  public static final PelagoProjectPushBelowUnpack INSTANCE = new PelagoProjectPushBelowUnpack();

  protected PelagoProjectPushBelowUnpack() {
    super(operand(Project.class, operand(PelagoUnpack.class, operand(PelagoTableScan.class, any()))));
  }

  public void onMatch(RelOptRuleCall call) {
    Project           project = (Project          ) call.rel(0);
    PelagoUnpack      unpack  = (PelagoUnpack     ) call.rel(1);
    PelagoTableScan   scan    = (PelagoTableScan  ) call.rel(2);


    final Mappings.TargetMapping mapping = project.getMapping();

    SortedSet<Integer> projects2 = new TreeSet();
    for (RexNode x: project.getChildExps()){
      RelOptUtil.InputReferencedVisitor vis = new RelOptUtil.InputReferencedVisitor();
      x.accept(vis);
      projects2.addAll(vis.inputPosReferenced);
    }

    int[] scanfields = scan.fields();
    int[] fields = new int[projects2.size()]; //getProjectFields(project.getProjects(), scan);
    int[] revfields = new int[scan.getRowType().getFieldCount()];
    int i = 0;
    for (Integer j: projects2){//int i = 0; i < projects2.size(); i++) {
      fields[i] = scanfields[j];
      revfields[j] = i - j;
      i++;
    }

    RelOptUtil.RexInputConverter conv = new RelOptUtil.RexInputConverter(scan.getCluster().getRexBuilder(), scan.getRowType().getFieldList(), revfields);
    List<RexNode> projs = new ArrayList<RexNode>();
    boolean isId = true;
    for (int j = 0 ; j < project.getProjects().size() ; ++j) {
      projs.add(project.getProjects().get(j).accept(conv));
      RexNode p = project.getProjects().get(j).accept(conv);
      isId = isId && (p instanceof RexInputRef && ((RexInputRef) p).getIndex() == j);
    }

    var nscan =
        PelagoTableScan.create(
            scan.getCluster(),
            scan.getTable(),
            scan.pelagoTable(),
            fields
        );

    var nnscan = call.getPlanner().register(nscan, null);

    RelNode in = call.getPlanner().register(PelagoUnpack.create(nnscan, RelPacking.UnPckd), null);

    if (!isId){
      in = project.copy(
        project.getTraitSet(),
        in,
        projs,
        project.getRowType()
      );
    }
    call.transformTo(in);//, ImmutableMap.of(project, in));
  }

//  @Override public boolean matches(final RelOptRuleCall call) {
//    if (!super.matches(call)) return false;
//    PelagoTableScan   scan    = (PelagoTableScan  ) call.rel(2);
//    return scan.getPluginInfo().get("type") == "block";
//  }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/rules/PelagoProjectPushBelowUnpack.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.rules

import ch.epfl.dias.calcite.adapter.pelago.rel.{PelagoTableScan, PelagoUnpack}
import ch.epfl.dias.calcite.adapter.pelago.traits.RelPacking
import org.apache.calcite.plan.RelOptRule.{any, operand}
import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelOptUtil}
import org.apache.calcite.rel.core.Project
import org.apache.calcite.rex.{RexInputRef, RexNode}

import java.util
import scala.collection.JavaConverters._

object PelagoProjectPushBelowUnpack {
  val INSTANCE = new PelagoProjectPushBelowUnpack
}

class PelagoProjectPushBelowUnpack protected ()
    extends RelOptRule(
      operand(
        classOf[Project],
        operand(classOf[PelagoUnpack], operand(classOf[PelagoTableScan], any))
      )
    ) {
  override def onMatch(call: RelOptRuleCall): Unit = {
    val project = call.rel(0).asInstanceOf[Project]
    val unpack = call.rel(1).asInstanceOf[PelagoUnpack]
    val scan = call.rel(2).asInstanceOf[PelagoTableScan]
    val mapping = project.getMapping
    val projects2 = new util.TreeSet[Integer]
    for (x <- project.getProjects.asScala) {
      val vis = new RelOptUtil.InputReferencedVisitor
      x.accept(vis)
      projects2.addAll(vis.inputPosReferenced)
    }
    val scanfields = scan.fields
    val fields =
      new Array[Int](
        projects2.size
      ) //getProjectFields(project.getProjects(), scan);
    val revfields = new Array[Int](scan.getRowType.getFieldCount)
    var i = 0
    for (j <- projects2.asScala) { //int i = 0; i < projects2.size(); i++) {
      fields(i) = scanfields(j)
      revfields(j) = i - j
      i += 1
    }
    val conv = new RelOptUtil.RexInputConverter(
      scan.getCluster.getRexBuilder,
      scan.getRowType.getFieldList,
      revfields
    )
    val projs = new util.ArrayList[RexNode]
    var isId = true
    for (j <- 0 until project.getProjects.size) {
      projs.add(project.getProjects.get(j).accept(conv))
      val p = project.getProjects.get(j).accept(conv)
      isId = isId && (p
        .isInstanceOf[RexInputRef] && p.asInstanceOf[RexInputRef].getIndex == j)
    }
    val nscan = PelagoTableScan.create(
      scan.getCluster,
      scan.getTable,
      scan.pelagoTable,
      fields
    )
    val nnscan = call.getPlanner.register(nscan, null)
    var in = call.getPlanner
      .register(PelagoUnpack.create(nnscan, RelPacking.UnPckd), null)
    if (!isId)
      in = project.copy(project.getTraitSet, in, projs, project.getRowType)
    call.transformTo(in) //, ImmutableMap.of(project, in));

  }
//  @Override public boolean matches(final RelOptRuleCall call) {
//    if (!super.matches(call)) return false;
//    PelagoTableScan   scan    = (PelagoTableScan  ) call.rel(2);
//    return scan.getPluginInfo().get("type") == "block";
//  }
}
